File: examples-objc/LocusLabsSamples/LocusLabsSamples/SampleVCs/DirectionsStepsETA.m
//
//  DirectionsStepsETA.m

//
//  Copyright © 2020 LocusLabs. All rights reserved.
//

#import "DirectionsStepsETA.h"
#import <LocusLabsSDK/LocusLabsSDK.h>

@interface DirectionsStepsETA () <LLMapViewDelegate, LLVenueDatabaseDelegate>

@property (nonatomic, weak)   LLMapView         *mapView;
@property (nonatomic, strong) LLVenue           *venue;
@property (nonatomic, strong) LLVenueDatabase   *venueDatabase;

@end

@implementation DirectionsStepsETA


#pragma mark Lifecycle

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    [LLLocusLabs setup].accountId = @"A11F4Y6SZRXH4X";

    // Create a new LLMapView, register as its delegate and add it as a subview
    LLMapView *mapView = [[LLMapView alloc] initWithFrame:self.view.bounds];
    mapView.autoresizingMask = (UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight);
    mapView.delegate = self;
    [self.view addSubview:mapView];
    
    self.mapView = mapView;
    self.mapView.showTopSafeAreaOverlay = NO;
    self.mapView.darkMode = self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark;
    
    // Get an instance of LLVenueDatabase, set it's mapview and register as its delegate
    self.venueDatabase = [LLVenueDatabase venueDatabaseWithMapView:self.mapView];
    self.venueDatabase.delegate = self;
    
    [self.venueDatabase loadVenueAndMap:@"lax" block:^(LLVenue *venue, LLMap *map, LLFloor *floor, LLMarker *marker) {
        
        self.venue = venue;
    }];
}

- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    
    [super traitCollectionDidChange:previousTraitCollection];
    
    self.mapView.darkMode = self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark;
}

#pragma mark Custom

- (void)getDirectionsAndETAWheelchairAccessibleOnly:(BOOL)wheelchairAccessibleOnly {
    
    // From: Blu20 Bar, Level 3, Terminal 6, ID 1025
    // To: Gate 73, Level 3, Terminal 7, ID 551
    // Note: if you dont have poi data to get a position from, you can also construct a LLPosition directly using the "initWithFloorId:latLng" method
    [self.venue.poiDatabase loadPois:@[@"1025", @"551"] withBlock:^(NSArray *pois) {
       
        LLPOI *startPOI = nil;
        LLPOI *endPOI = nil;
        for (LLPOI *poi in pois) {
            
            if ([poi.poiId isEqualToString:@"1025"]) startPOI = poi;
            else if ([poi.poiId isEqualToString:@"551"]) endPOI = poi;
        }
        
        LLDirectionsRequest *directionsRequest = [[LLDirectionsRequest alloc] initWithStartPosition:startPOI.position endPosition:endPOI.position];
        directionsRequest.forceWheelchairAccessibleRoute = wheelchairAccessibleOnly;
        
        [self.venue navigate:directionsRequest completion:^(LLNavigationPath *navigationPath, NSError *error) {
            
            NSLog(@"ETA:%@", navigationPath.eta);
            NSArray *wayPoints = navigationPath.waypoints;
           
            for (LLWaypoint *wayPoint in wayPoints) {
                
                NSLog(@"Waypoint:");
                NSLog(@"ETA:%@", wayPoint.eta);
                NSLog(@"Distance:%@", wayPoint.distance);
                NSLog(@"Details:%@", wayPoint.details);
                NSLog(@"Security Checkpoint?:%@", wayPoint.securityCheckpoint);
                NSLog(@"\n");
            }
        }];
    }];
}

#pragma mark Delegates - LLVenueDatabase

- (void)venueDatabase:(LLVenueDatabase *)venueDatabase venueLoadFailed:(NSString *)venueId code:(LLDownloaderError)errorCode message:(NSString *)message {

    // Handle failures here
}

#pragma mark Delegates - LLMapView

- (void)mapViewDidClickBack:(LLMapView *)mapView {
    
    // The user tapped the "Cancel" button while the map was loading. Dismiss the app or take other appropriate action here
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)mapViewReady:(LLMapView *)mapView {
    
    [self getDirectionsAndETAWheelchairAccessibleOnly:NO];
}

- (UIViewController *)presentingControllerForMapView:(LLMapView *)mapView forContext:(LLMapViewPresentationContext)context {
    
    // Return a viewcontroller the SDK can use to present alerts
    return self;
}

@end

File: LocusLabsSamplesSwift/LocusLabsSamplesSwift/Samples/DirectionsStepsETA.swift

//  Copyright © 2021 Atrius, part of Acuity Brands. All rights reserved.
//

import UIKit

// NB - You may need to run "pod update" before running this code

class DirectionsStepsETA: UIViewController, LLVenueDatabaseDelegate, LLMapViewDelegate {

    // Vars
    var venueDatabase:      LLVenueDatabase!
    var venue:              LLVenue?
    var mapView:            LLMapView?
    
    // MARK: Lifecycle
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        // Initialize the SDK with the accountId provided
        LLLocusLabs.setup().accountId = "A11F4Y6SZRXH4X"
        
        // Create a new LLMapView, register as its delegate and add it as a subview
        mapView = LLMapView(frame: view.bounds)
        mapView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView?.delegate = self
        view.addSubview(mapView!)
        
        mapView?.showTopSafeAreaOverlay = false
        mapView?.darkMode = traitCollection.userInterfaceStyle == .dark
        
        // Get an instance of LLVenueDatabase, register as its delegate and load the venue LAX
        venueDatabase = LLVenueDatabase(mapView: mapView)
        venueDatabase.delegate = self
        
        // Load the venue LAX async
        venueDatabase.loadVenueAndMap("lax") { (_venue: LLVenue?, _map: LLMap?, _floor: LLFloor?, _marker: LLMarker?) in
            
            self.venue = _venue
        }
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        
        super.traitCollectionDidChange(previousTraitCollection)
        
        mapView?.darkMode = traitCollection.userInterfaceStyle == .dark
    }
    
    // MARK: Custom
    func getDirectionsAndETA(wheelChairOnly: Bool) {
        
        // From: Blu20 Bar, Level 3, Terminal 6, ID 1025
        // To: Gate 73, Level 3, Terminal 7, ID 551
        // Note: if you dont have poi data to get a position from, you can also construct a LLPosition directly using this method: "LLPosition *positionStart = [[LLPosition alloc] initWithFloorId: latLng:];
        venue?.poiDatabase()?.loadPois(["1025", "551"], with: { pois in
        
            var startPOI: LLPOI? = nil
            var endPOI: LLPOI? = nil
            for poi in pois as! [LLPOI] {
                
                if poi.poiId == "1025" {startPOI = poi}
                else if poi.poiId == "551" {endPOI = poi}
            }
            
            if startPOI == nil || endPOI == nil {return}
            
            let directionsRequest = LLDirectionsRequest(start: startPOI!.position!, end: endPOI!.position!)
            directionsRequest.forceWheelchairAccessibleRoute = wheelChairOnly
            
            self.venue?.navigate(directionsRequest, completion: { (navigationPath, error) in
                
                print("ETA:", navigationPath?.eta ?? 0)
                for wayPoint in navigationPath?.waypoints as! [LLWaypoint] {
                    
                    print("Waypoint:");
                    print("ETA:%@", wayPoint.eta ?? 0);
                    print("Distance:%@", wayPoint.distance ?? 0);
                    print("Details:%@", wayPoint.details ?? "");
                    print("Security Checkpoint?:%@", wayPoint.securityCheckpoint ?? false);
                    print("\n");
                }
            })
        })
    }
    
    // MARK: Delegates - LLVenueDatabase
    func venueDatabase(_ venueDatabase: LLVenueDatabase!, venueLoadFailed venueId: String!, code errorCode: LLDownloaderError, message: String!) {
        
        // Handle failures here
    }
    
    // MARK: Delegates - LLMapView
    func mapViewDidClickBack(_ mapView: LLMapView!) {
        
        // The user tapped the "Cancel" button while the map was loading. Dismiss the app or take other appropriate action here
        navigationController?.popViewController(animated: true)
    }
    
    func mapViewReady(_ mapView: LLMapView!) {
        
        getDirectionsAndETA(wheelChairOnly: false)
    }
    
    func presentingController(for mapView: LLMapView!, for context: LLMapViewPresentationContext) -> UIViewController! {
        
        // Return a viewcontroller the SDK can use to present alerts
        return self
    }
}

